### DISCUSSION TITLE
SFTPGo Docker container is much smaller than the Joplin Server container!

### DISCUSSION THREAD

**Discussion by derekmahar** [drakkan/sftpgo#865] [General]
Here is an interesting SFTPGo anecdote and endorsement.

I found SFTPGo while searching for a WebDAV server to run on my Raspberry Pi 4 home server.  My primary motivation for installing a WebDAV server was to synchronize my [Joplin](https://joplinapp.org/) notes.  Joplin can synchronize notes to [Joplin Server](https://github.com/laurent22/joplin/blob/dev/readme/changelog_server.md), but the uncompressed [Joplin Server Docker image for arm64](https://hub.docker.com/r/etechonomy/joplin-server) for the Raspberry Pi 4 is 1.24 GB!  Compare this with the uncompressed ["alpine-slim"](https://github.com/drakkan/sftpgo/blob/v2.3.0/Dockerfile.alpine) [SFTPGo Docker container image](https://hub.docker.com/r/drakkan/sftpgo) for [arm64](https://hub.docker.com/layers/drakkan/sftpgo/2.3.x-alpine-slim/images/sha256-5515bde5c2ca3f3edd921ef11eb0178961b2ddba39d768770b5b1cd8357a8b3f?context=explore) which is only 46.5 MB!  Why would anyone use Joplin Server instead of SFTPGo or some other small WebDAV server implementation?

```
ubuntu@raspberrypi1:~$ docker image ls etechonomy/joplin-server
REPOSITORY                 TAG       IMAGE ID       CREATED        SIZE
etechonomy/joplin-server   latest    7120b5abbe94   4 months ago   1.24GB
ubuntu@raspberrypi1:~$ docker image ls drakkan/sftpgo
REPOSITORY       TAG           IMAGE ID       CREATED      SIZE
drakkan/sftpgo   alpine-slim   404b0f1a1261   3 days ago   46.5MB
```

**Comment by drakkan** [comment]
Thanks for posting this, the distroless based image should also be few MB smaller than the alpine one.
SFTPGo is written in Go so we can deploy static or semi-static (only a libc is required) binaries, we don't need system dependencies. 

Honestly I don't know joplin (this is my first time hearing about it), but it seems to have more features than just a webdav server.

Glad you like SFTPGo, I'm doing my best to make it useful

**Reply by derekmahar** [reply]
> Thanks for posting this, the distroless based image should also be few MB smaller than the alpine one.

Which tag is the distroless image?

> SFTPGo is written in Go so we can deploy static or semi-static (only a libc is required) binaries, we don't need system dependencies.

Why did you choose Go over some other programming language like Java or Rust?  (I seem to be continuously exploring the programming language space.  There are so many choices!)
 
> Honestly I don't know joplin (this is my first time hearing about it), but it seems to have more features than just a webdav server.

Joplin itself is a note taking application that supports Markdown, notebooks, and tags.  Joplin Server is a related project whose only purpose (as far as I can tell) is to share notes.  The only synchronization facility available in Joplin is WebDAV, so I think Joplin Server supports only this protocol.

> Glad you like SFTPGo, I'm doing my best to make it useful

Keep up the good work!

**Reply by drakkan** [reply]
> > Thanks for posting this, the distroless based image should also be few MB smaller than the alpine one.
> 
> Which tag is the distroless image?
> 

`distroless-slim` or `edge-distroless-slim`. Take a look [here](https://github.com/drakkan/sftpgo/tree/main/docker) for more details

> > SFTPGo is written in Go so we can deploy static or semi-static (only a libc is required) binaries, we don't need system dependencies.
> 
> Why did you choose Go over some other programming language like Java or Rust? (I seem to be continuously exploring the programming language space. There are so many choices!)
> 

SFTPGo started out as a hobby project, and it's still a hobby project even though some companies are starting to sponsor me and I know very large companies that use it (but sadly large companies don't like to sponsor, but this is another story ...).

For a hobby project, java is out of the question. Sometimes I need to use it for my daily work and I hate it.

I wanted to learn a new programming language and the candidates were Go and Rust.
I decided to use Go because it has a quite mature SFTP library (that I now co-maintain). Initially only SFTP and local filesystem was supported and I never imagined adding WebDAV, FTP or cloud storage backends.

I have yet to find time to learn Rust :smile: 

I hope that SFTPGo can become more of a hobby project in the future and that I can dedicate more time to it, we'll see what happens.

> > Honestly I don't know joplin (this is my first time hearing about it), but it seems to have more features than just a webdav server.
> 
> Joplin itself is a note taking application that supports Markdown, notebooks, and tags. Joplin Server is a related projected whose only purpose (as far as I can tell) is to share notes. The only synchronization facility available in Joplin is WebDAV, so I think Joplin Server supports only this protocol.
> 
> > Glad you like SFTPGo, I'm doing my best to make it useful
> 
> Keep up the good work!

**Reply by derekmahar** [reply]
> `distroless-slim` or `edge-distroless-slim`. Take a look [here](https://github.com/drakkan/sftpgo/tree/main/docker) for more details

```
ubuntu@raspberrypi1:~$ docker image ls drakkan/sftpgo
REPOSITORY       TAG               IMAGE ID       CREATED      SIZE
drakkan/sftpgo   alpine-slim       404b0f1a1261   3 days ago   46.5MB
drakkan/sftpgo   distroless-slim   5b7ec143b239   3 days ago   40.5MB
```

**Reply by derekmahar** [reply]
> SFTPGo started out as a hobby project, and it's still a hobby project even though some companies are starting to sponsor me and I know very large companies that use it (but sadly large companies don't like to sponsor, but this is another story ...).

Have you approached these larger companies for sponsorship?

> For a hobby project, java is out of the question. Sometimes I need to use it for my daily work and I hate it.

I've spent most of my career (since 2002) as a Java developer.  Before that, I was a C++ developer.  I don't hate Java, but I don't like it as much as I did before.  When it was first released, it had a lot of novel features like a virtual machine and garbage collector that weren't available in C++.  These features are still its biggest strengths, but in my opinion, Java suffers from complex class hierarchies and dependency injection frameworks that are difficult to use, awkward functional programming syntax, excessive annotation magic, dreaded null pointer exceptions.  I learned C++ and Java when object-oriented design and programming were very popular and growing.  Since then, OOP has lost some popularity to functional programming.  Scala offers both, but with additional complexity.

> I wanted to learn a new programming language and the candidates were Go and Rust.

I was facing the same choice a few years ago, but chose Rust instead because it didn't have null pointers, but did have generics which Go didn't have until [recently](https://go.dev/blog/go1.18).  However, Rust is arguably more difficult than Go, so I still have much more to learn and need more experience which will take practise.

> I decided to use Go because it has a quite mature SFTP library (that I now co-maintain). Initially only SFTP and local filesystem was supported and I never imagined adding WebDAV, FTP or cloud storage backends.

These are welcome additions!  Go seems to be very popular and has many useful networking libraries which reinforces its popularity in a continuous feedback loop.  Rust and Java are in a similar position.  I wish that it were much easier to share libraries between programming languages.  Microsoft had the right idea with .NET, but it hasn't caught on much in the open source community.

> I have yet to find time to learn Rust 😄

I haven't spent enough time with Rust, but it's one of my priorities, as is learning more Haskell.

> I hope that SFTPGo can become more of a hobby project in the future and that I can dedicate more time to it, we'll see what happens.

Best of luck!  It's a worthwhile project!  I appreciate SFTPGo because it's small, easy to install and launch, but still highly configurable with sensible defaults, has a basic, but adequate Web client, and is quite responsive on my Raspberry Pi 4.  I think SFTPGo would be a highly welcome addition to the [LinuxServer](https://www.linuxserver.io/), [DockSTARTer](https://dockstarter.com/), and [SWAG](https://github.com/linuxserver/docker-swag) projects.  It wouldn't require much effort to add two NGINX reverse proxy configuration files to SWAG, for example.
